### PR TITLE
[APPC-1161] Implemented pre selected payment methods

### PR DESCRIPTION
**Pull request by carvalho2707**
**What does this PR do?**

   This PR introduces a feature to have pre selected payment methods. There was also a change to payments's flow where the wallet addres(footer) was removed.

**Database changed?**

   No

**Where should the reviewer start?**

- [ ] PaymentMethodsFragment.kt

**How should this be manually tested?**

Until the fix for the orderreferenceid is fixed , change the field orderReference in the RemoteRepository to null.

  Flow on how to test this or QA Tickets related to this use-case: [APPC-1161](https://aptoide.atlassian.net/browse/APPC-1161)

**What are the relevant tickets?**

  Tickets related to this pull-request: [APPC-1161](https://aptoide.atlassian.net/browse/APPC-1161)

**Questions:**

   Does this add new dependencies which need to be added? (Eg. new libs, new keys, etc.)

**Code Review Checklist**

- [ ] Architecture
- [ ] Documentation on public interfaces
- [ ] Database changed?
- [ ] If yes - Database migration?
- [ ] Remove comments & unused code & forgotten testing Logs
- [ ] Codestyle
- [ ] Functional QA tests pass

### CODE REVIEW

File: app/src/main/java/com/asfoundation/wallet/billing/BillingService.kt [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  fun getTransactionUid(): String?
+
+  fun getAuthorization(productName: String?,
+                       developerAddress: String?, payload: String?, origin: String,
+                       priceValue: BigDecimal,
+                       priceCurrency: String, type: String, callback: String?,
+                       orderReference: String?,
+                       appPackageName: String): Observable<AdyenAuthorization>
+
+  fun getAuthorization(origin: String, priceValue: BigDecimal,
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/billing/BillingService.kt]: These 2 methods can be better splitted (a line can take more arguments before it breaks to another line)

File: app/src/main/java/com/asfoundation/wallet/billing/adyen/Adyen.java [hunk 2]
```diff
@@ -169,6 +169,12 @@ private void cancelPayment() {
     paymentRequest = null;
   }
 
+  public Completable deletePaymentMethod(PaymentMethod paymentMethod) {
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/billing/adyen/Adyen.java]: paymentMethod is not being used

File: app/src/main/java/com/asfoundation/wallet/ui/iab/AdyenAuthorizationFragment.java [hunk 3]
```diff
@@ -105,16 +109,22 @@
   private PublishRelay<Boolean> keyboardBuyRelay;
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/iab/AdyenAuthorizationFragment.java:110]: walletAddressFooter is not being used

File: app/src/main/java/com/asfoundation/wallet/ui/iab/AdyenAuthorizationFragment.java [hunk 4]
```diff
@@ -182,14 +198,26 @@ public static AdyenAuthorizationFragment newInstance(String skuId, String type,
...
+        .getParent();
+    lp = (ViewGroup.MarginLayoutParams) expirationParent.getLayoutParams();
+    lp.setMargins(0, marginTop, 0, 0);
+    expirationParent.setLayoutParams(lp);
+
+    View cvvParent = (View) cardForm.findViewById(R.id.bt_card_form_cvv)
+        .getParent();
+    lp = (ViewGroup.MarginLayoutParams) cvvParent.getLayoutParams();
+    lp.setMargins(0, marginTop, 0, 0);
+    cvvParent.setLayoutParams(lp);
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/iab/AdyenAuthorizationFragment.java]: This could be extracted to a private method, so that in the future people understand better what it does (despite the comment) and to improve readability of the onViewCreated method

File: app/src/main/java/com/asfoundation/wallet/ui/iab/AdyenAuthorizationPresenter.java [hunk 5]
```diff
@@ -128,15 +120,29 @@ public void present(@Nullable Bundle savedInstanceState) {
...
+    handleMorePaymentMethodClicks();
+  }
+
+  private void handleMorePaymentMethodClicks() {
+    if (isPreSelected) {
+      disposables.add(view.getMorePaymentMethodsClicks()
+          .observeOn(viewScheduler)
+          .doOnNext(o -> {
+            inAppPurchaseInteractor.removePreSelectedPaymentMethod();
+            showMoreMethods();
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/iab/AdyenAuthorizationPresenter.java]: why not just view.showMoreMethods()

File: app/src/main/java/com/asfoundation/wallet/ui/iab/InAppPurchaseInteractor.java [hunk 6]
```diff
@@ -261,12 +266,15 @@ private boolean shouldSwap(List<PaymentMethod> paymentMethods, int position) {
...
-          && paymentMethod.getAvailability() != null
-          && paymentMethod.getAvailability()
-          .equals("UNAVAILABLE")) {
-        iterator.remove();
+      } else {
+        if (paymentMethod.getGateway()
+            .getName() == (Gateway.Name.myappcoins)) {
+          paymentMethod.getAvailability();
+          if (paymentMethod.getAvailability()
+              .equals("UNAVAILABLE")) {
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/iab/InAppPurchaseInteractor.java]: Why did you change the logic here?

File: app/src/main/java/com/asfoundation/wallet/ui/iab/InAppPurchaseInteractor.java [hunk 7]
```diff
@@ -284,4 +292,31 @@ private PaymentMethod mapPaymentMethods(PaymentMethodEntity paymentMethod,
...
+
+  String getPreSelectedPaymentMethod() {
+    return sharedPreferences.getString(PRE_SELECTED_PAYMENT_METHOD_KEY,
+        PaymentMethodsView.PaymentMethodId.APPC_CREDITS.getId());
+  }
+
+  void savePreSelectedPaymentMethod(String paymentMethod) {
+    SharedPreferences.Editor editor = sharedPreferences.edit();
+    editor.putString(PRE_SELECTED_PAYMENT_METHOD_KEY, paymentMethod);
+    editor.putString(LAST_USED_PAYMENT_METHOD_KEY, paymentMethod);
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/iab/InAppPurchaseInteractor.java:356]: Isn’t the last always the same as the pre-selected?

File: app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsFragment.java [hunk 8]
```diff
@@ -68,18 +71,17 @@
...
   private final Map<String, Bitmap> loadedBitmaps = new HashMap<>();
-  PaymentMethodsPresenter presenter;
   @Inject InAppPurchaseInteractor inAppPurchaseInteractor;
   @Inject BillingAnalytics analytics;
   @Inject BdsPendingTransactionService bdsPendingTransactionService;
   @Inject Billing billing;
-  @Inject WalletService walletService;
   @Inject GamificationInteractor gamification;
   @Inject PaymentMethodsMapper paymentMethodsMapper;
+  @Inject BillingFactory billingFactory;
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsFragment.java]: Not used

File: app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsPresenter.kt [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+    private val billing: Billing,
+    private val analytics: BillingAnalytics,
+    private val isBds: Boolean,
+    private val developerPayload: String,
+    private val uri: String,
+    private val gamification: GamificationInteractor,
+    private val transaction: TransactionBuilder,
+    private val paymentMethodsMapper: PaymentMethodsMapper) {
+
+  private var transactionValue: Double = 0.toDouble()
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsPresenter.kt]: This can passed as an argument of the presenter

**Review by JDSM01 — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/asfoundation/wallet/ui/iab/AdyenAuthorizationPresenter.java (hunk 5 above)
Comment (carvalho2707) [reply app/src/main/java/com/asfoundation/wallet/ui/iab/AdyenAuthorizationPresenter.java]: Because it means we don't have to have a pre selected payment method anymore.

File: app/src/main/java/com/asfoundation/wallet/ui/iab/InAppPurchaseInteractor.java (hunk 6 above)
Comment (carvalho2707) [reply app/src/main/java/com/asfoundation/wallet/ui/iab/InAppPurchaseInteractor.java]: Because paymentMethod.getAvailability() returns non null values always and thus the check is unnecessary

File: app/src/main/java/com/asfoundation/wallet/ui/iab/InAppPurchaseInteractor.java (hunk 7 above)
Comment (carvalho2707) [reply app/src/main/java/com/asfoundation/wallet/ui/iab/InAppPurchaseInteractor.java:356]: PRE_SELECTED_PAYMENT_METHOD_KEY is for the new flow where we don't show the payment methods list.
LAST_USED_PAYMENT_METHOD_KEY is used when there is no pre selected payment method but we want to have the radio check on the last used (before clicking on "more payment method)

File: app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsPresenter.kt [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+    for (paymentMethod in paymentMethods) {
+      if (paymentMethod.id == preSelectedPreference) {
+        return paymentMethod
+      }
+    }
+    return null
+  }
+
+  private fun getLastUsedPaymentMethod(paymentMethods: List<PaymentMethod>): String {
+    val preSelectedPreference = inAppPurchaseInteractor.lastUsedPaymentMethod
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsPresenter.kt]: variable name

File: app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsPresenter.kt [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+  }
+
+  private fun handleErrorDismisses() {
+    disposables.add(Observable.merge(view.errorDismisses(), view.onBackPressed())
+        .flatMapCompletable { itemAlreadyOwned ->
+          if (itemAlreadyOwned) {
+            return@flatMapCompletable getPurchases(transaction.skuId).doOnNext { purchase ->
+              finish(purchase, true)
+            }
+                .ignoreElements()
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsPresenter.kt:295]: Use “it” instead of purchase. Remove line break of ignore elements

File: app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsPresenter.kt [hunk 12]
```diff
@@ -0,0 +1,337 @@
...
+          } else {
+            return@flatMapCompletable Completable.fromAction { Action { this.close() } }
+          }
+        }
+        .subscribe({ }, { this.showError(it) }))
+  }
+
+  private fun finish(purchase: Purchase, itemAlreadyOwned: Boolean?) {
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsPresenter.kt]: itemAlreadyOwned is never null

**Review by JDSM01 — CHANGES_REQUESTED** [review]

File: app/src/main/res/layout/selected_payment_method_cc.xml [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+      android:layout_width="0dp"
+      android:layout_height="wrap_content"
+      android:layout_below="@id/fragment_credit_card_authorization_pre_authorized_card"
+      android:layout_marginStart="16dp"
+      android:layout_marginTop="12dp"
+      android:layout_marginBottom="12dp"
+      android:checked="true"
+      android:text="@string/dialog_credit_card_remember"
+      android:textSize="11sp"
+      android:visibility="visible"
```
Comment (JDSM01) [inline app/src/main/res/layout/selected_payment_method_cc.xml]: not needed

File: app/src/main/res/layout-land/dialog_credit_card_authorization_pre_selected.xml [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+      android:layout_gravity="center"
+      android:animateLayoutChanges="true"
+      >
+    <ProgressBar
+        android:id="@+id/fragment_credit_card_authorization_progress_bar"
+        android:layout_width="80dp"
+        android:layout_height="80dp"
+        android:layout_centerInParent="true"
+        android:indeterminateDrawable="@drawable/gradient_progress"
+        android:visibility="visible"
```
Comment (JDSM01) [inline app/src/main/res/layout-land/dialog_credit_card_authorization_pre_selected.xml]: not needed

**Review by JDSM01 — CHANGES_REQUESTED** [review]

File: app/src/main/res/layout-land/payment_methods_layout.xml [hunk 15]
```diff
@@ -278,8 +184,38 @@
...
+          app:layout_constraintBottom_toBottomOf="@id/bottom_separator_buttons"
+          app:layout_constraintEnd_toEndOf="parent"
+          app:layout_constraintStart_toStartOf="@id/mid_guideline"
+          app:layout_constraintTop_toTopOf="@id/top_separator"
+          app:layout_constraintVertical_bias="0.3"
+          />
+
+      <TextView
+          android:id="@+id/more_payment_methods"
+          style="@style/TextAppearance.AppCompat.Body1.Roboto.Medium"
```
Comment (JDSM01) [inline app/src/main/res/layout-land/payment_methods_layout.xml:202]: Text color doesn’t match the one in zeplin

File: app/src/main/res/layout/selected_payment_method.xml [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+      android:layout_marginBottom="24dp"
+      android:scaleType="fitCenter"
+      app:layout_constraintBottom_toBottomOf="parent"
+      app:layout_constraintStart_toStartOf="parent"
+      app:layout_constraintTop_toTopOf="parent"
+      tools:src="@drawable/ic_payment_method_paypal"
+      />
+
+  <TextView
+      android:id="@+id/payment_method_description"
```
Comment (JDSM01) [inline app/src/main/res/layout/selected_payment_method.xml]: Text is not centered in some cases in portrait (e.g. PayPal)

File: app/src/main/res/layout/selected_payment_method.xml [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+      android:layout_marginTop="16dp"
+      android:textSize="16sp"
+      app:layout_constraintBottom_toTopOf="@id/payment_method_secondary"
+      app:layout_constraintStart_toEndOf="@id/payment_method_ic"
+      app:layout_constraintTop_toTopOf="@id/payment_method_ic"
+      tools:layout_editor_absoluteY="116dp"
+      tools:text="Paypal"
+      />
+  <TextView
+      android:id="@+id/payment_method_secondary"
```
Comment (JDSM01) [inline app/src/main/res/layout/selected_payment_method.xml:56]: This could be gone by default

File: app/src/main/java/com/asfoundation/wallet/ui/iab/IabActivity.kt [hunk 18]
```diff
@@ -0,0 +1,253 @@
...
+    const val TRANSACTION_AMOUNT = "transaction_amount"
+    const val TRANSACTION_CURRENCY = "transaction_currency"
+    const val DEVELOPER_PAYLOAD = "developer_payload"
+    const val BDS = "BDS"
+    const val WEB_VIEW_REQUEST_CODE = 1234
+    const val IS_BDS_EXTRA = "is_bds_extra"
+
+    @JvmStatic
+    fun newIntent(activity: Activity, previousIntent: Intent, transaction: TransactionBuilder,
+                  isBds: Boolean?, developerPayload: String): Intent {
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/iab/IabActivity.kt]: developerPayload can be null

**Review by JDSM01 — CHANGES_REQUESTED** [review]
Margin from header is to far on the left
Ask a friend not working
PaymentMethodView -> rotate -> buy with credits -> previous view doesn’t disappear
Landscape margin bottom in pre-selected view is too large
Portrait margin bottom in pre-selected view is too short
Pre-selectedView -> rotate-> rotate -> paymentMethodView appears in the background
Going from a  payment with bonus and a payment without bonus makes view shrink/enlarge
Use case missing: buy with credits --> no longer have credits --> should show payment list view
Item already owned not working

**Review by JDSM01 — CHANGES_REQUESTED** [review]
- [x] Bonus landscape layout is still botched

- [x] Payment List layout doesn’t match the one in the merged appcoins branch. (Space to the top, payment_methods_layout)

- [x] Bottom list line (top bonus line) doesn’t appear in portrait ( if preselected and then click on more)

- [x] ProductName in IabActivity can be null

- [x] Item already owned nao tem listener no botao de erro

- [x] Margin between bonus layout and bottom separator(botto top line)

**Review by JDSM01 — APPROVED** [review]

**Review by JDSM01 — CHANGES_REQUESTED** [review]
Wrong flow: Made a buy with credit card, mock response to remove CC from response, Merge Appcoins appears as pre-selected

File: app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsFragment.java [hunk 19]
```diff
@@ -211,7 +211,8 @@ public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup c
     preSelectedMethodView = view.findViewById(R.id.layout_pre_selected);
     preSelectedIcon = preSelectedMethodView.findViewById(R.id.payment_method_ic);
     preSelectedName = preSelectedMethodView.findViewById(R.id.payment_method_description);
-    pareSelectedNameSingle = preSelectedMethodView.findViewById(R.id.payment_method_description_single);
+    pareSelectedNameSingle =
```
Comment (JDSM01) [inline app/src/main/java/com/asfoundation/wallet/ui/iab/PaymentMethodsFragment.java]: name

**Review by JDSM01 — CHANGES_REQUESTED** [review]

**Review by JDSM01 — CHANGES_REQUESTED** [review]
- [x] Local Payment asynchronous

- [x] Buy button doesn’t work on pre-selected

- [x] Buy on CC without cvc gives an error that ends the flow

**Review by JDSM01 — APPROVED** [review]
